azure/test.yml

```diff
@@ -24,33 +25,38 @@ jobs:
           vmImage: macOS-10.14
         Windows:
           vmImage: windows-2019
-  ${{ if ne('true', parameters.cross) }}:
-    variables:
+  variables:
+    ${{ if ne('true', parameters.cross) }}:
       vmImage: ubuntu-16.04
+    options: ''
```

Comment:
I like this construction, though I'm a little worried about bash being "helpful" in splitting arguments and parsing quoted arguments incorrectly when we do it this way. Specifically because `echo` will get very confused if passed, say, an argument that contains a `'` inside of the `'` you surround its argument with.

Reply:
I agree, but I'm unsure of how else to create an empty string in pipelines.

Reply:
Ah, sorry, I wasn't referring so much to the empty string, but rather to if a user passes, say:
```yaml
parameters:
  features: "foo,'bar baz'"
```
That will currently make `echo` all sorts of confused since the first `'` in there will close the surrounding `'`. Of course maybe the answer here is just to punt on this entirely, but I _do_ think it's important if we introduce raw_opts into `$(options)`, since that is pretty likely to contain single quotes. It might be sufficient to document that it is not allowed to have single quotes, but even that restriction can be hard to work around. This is why the old code was written the way it was rather than constructing a parameter list at runtime :)

Reply:
I agree we should document it, but I think even with this rough edge the benefits of:

* Not having to multiply bool checks when adding a new option
* The ability to feed cargo exactly which args are required

outweighs it. YMMV 🤷‍♂

Reply:
Yeah, that's a good point. Here's my suggestion, based on the fact that `raw_opts` is most likely to cause trouble. Build up _two_ lists, one for cargo args and one for libtest args. Default `raw_opts` to `--`. Then, for `cargo test`, use:
```
- bash: cargo test $(cargo_opts) ${{ parameters.raw_opts }} $(libtest_opts)
```
and the same for the version with `--ignored`:
```
- bash: cargo test $(cargo_opts) ${{ parameters.raw_opts }} $(libtest_opts) --ignored
```
This way `raw_opts` never goes through `echo ''`, and thus quoting shouldn't be an issue, but we still avoid having to multiply the bool checks and retain the ability to provide specific cargo args.

What do you think?

Reply:
I think that solves the issues, I'll give it a bash now, but it might have to wait til tomorrow. Good idea!

Reply:
No worries at all -- there's no rush. We'd rather get it right!
I'm heading on vacation later today, so _may_ take a little longer to get back to you for later changes, just FYI :)

Reply:
No worries at all -- there's no rush. We'd rather get it right!
I'm heading on vacation later today, so _may_ take a little longer to get back to you for later changes, just FYI :)

Reply:
enjoy